### PR TITLE
81x MET XY correction cond DB

### PR DESCRIPTION
**Pull request by d4space**
MET XY Shift correction DB system:
Solving the problem of the closed "14923":
https://github.com/cms-sw/cmssw/pull/14923
which has the problem of "schema change" from the previous 80X version.

### CODE REVIEW

**Comment by cvuosalo** [comment]
@cmsbuild please test

File: JetMETCorrections/Modules/plugins/METCorrectorDBWriter.cc [hunk 1]
```diff
@@ -24,49 +27,53 @@ class  METCorrectorDBWriter : public edm::EDAnalyzer
...
 	      isect != isectend; ++isect ) {
-	  payload->push_back( i, METCorrectorParameters(fip.fullPath(),*isect), ilev + "_" + *isect );	  
-	  std::cout << "Added " << ilev + "_" + *isect <<  " to record " << i << std::endl;
+	  payload->push_back( ilev, METCorrectorParameters(fip.fullPath(),*isect), *isect );	  
+	  std::cout << "Added level " << levelName  + "_" + *isect <<  " to record "<<ilev<< std::endl;
 	}
       }
-      std::cout << "Added record " << i << std::endl;
-    }else{
+      std::cout << "Added record " << ilev << std::endl;
```
Comment (cvuosalo) [inline JetMETCorrections/Modules/plugins/METCorrectorDBWriter.cc]: Here `cout` is used, while on line 46 you use `LogDebug`. Is this code used in any workflow, or is it only called by a user?

Comment (d4space) [inline JetMETCorrections/Modules/plugins/METCorrectorDBWriter.cc]: It is called by a user only to make DB locally.

File: CondFormats/JetMETObjects/interface/METCorrectorParameters.h [hunk 2]
```diff
@@ -58,13 +59,15 @@ class METCorrectorParameters
...
         unsigned nParameters()              const {return mParameters.size();         }
+	std::string MetAxis()  const {return mMetAxis;       }
         int operator< (const Record& other) const {return xMin(0) < other.xMin(0);    }
       private:
         //-------- Member variables ----------
         unsigned           mNvar;
         std::vector<float> mMin;
         std::vector<float> mMax;
         std::vector<float> mParameters;
+	std::string        mMetAxis;
```
Comment (ggovi) [inline CondFormats/JetMETObjects/interface/METCorrectorParameters.h]: This data member addition is a schema change... To be clear, whenever a change affects directly or indirectly a top level class, you should create a new class.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
@cmsbuild please test
Jenkins DQM tests didn't run last time.

**Comment by cvuosalo** [comment]
+1

For #16131 429492c9c053f80da5d0e33c33b1b4bd99165d4a

Implementing MET X-Y shift correction from cond DB.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_8_1_X_2016-10-21-1100 show no significant differences.

**Comment by ggovi** [comment]
+1

File: CondFormats/JetMETObjects/src/MEtXYcorrectParameters.cc [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+#include <iterator>
+//#include <string>
+
+//------------------------------------------------------------------------ 
+//--- MEtXYcorrectParameters::Definitions constructor --------------------
+//--- takes specific arguments for the member variables ------------------
+//------------------------------------------------------------------------
+MEtXYcorrectParameters::Definitions::Definitions(const std::vector<std::string>& fBinVar, const std::vector<std::string>& fParVar, const std::string& fFormula )
+{
+  for(unsigned i=0;i<fBinVar.size();i++)
```
Comment (davidlange6) [inline CondFormats/JetMETObjects/src/MEtXYcorrectParameters.cc]: add a mBinVar.reserve(fBinVar.size())

File: CondFormats/JetMETObjects/interface/MEtXYcorrectParameters.h [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  // that are aware of all three collections. 
+  void validKeys(std::vector<key_type> & keys ) const;
+
+
+  // Helper method to find all of the sections in a given 
+  // parameters file
+  static void getSections( std::string inputFile,
+			   std::vector<std::string> & outputs );
+
+  static key_type getShiftMcFlavBin( std::string const & Flav );
```
Comment (davidlange6) [inline CondFormats/JetMETObjects/interface/MEtXYcorrectParameters.h]: Hi @d4space - why are these static? they reference member data, so they should not be

File: CondFormats/JetMETObjects/src/MEtXYcorrectParameters.cc [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+      handleError("MEtXYcorrectParameters::Definitions",sserr.str());
+    }
+    // No. of Bin Variable
+    LogDebug ("default")<<"Definitions===========";
+    ptclType = getSigned(tokens[0]);
+    unsigned nBinVar = getUnsigned(tokens[1]);
+    unsigned nParVar = getUnsigned(tokens[nBinVar+2]);
+    for(unsigned i=0;i<nBinVar;i++)
+    {
+      mBinVar.push_back(tokens[i+2]);
```
Comment (davidlange6) [inline CondFormats/JetMETObjects/src/MEtXYcorrectParameters.cc]: same comment here

File: JetMETCorrections/Type1MET/python/correctedPatMet_cff.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import FWCore.ParameterSet.Config as cms
+
+##____________________________________________________________________________||
+#labels used for the phi correction
+
+
+##____________________________________________________________________________||
+patMetTxy = cms.EDProducer(
```
Comment (d4space) [inline JetMETCorrections/Type1MET/python/correctedPatMet_cff.py]: i'm planing to add more

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #16131 c59cf4242d551f895757ecd7d8f66e222f6ae697

Second approval, after minor code revision, which is satisfactory. New Jenkins results are OK.

**Comment by ggovi** [comment]
+1

**Pull request merged**